manual/src/tutorials/memorymodel.etex

```diff
@@ -0,0 +1,790 @@
+
+The aim of the OCaml relaxed memory model is to precisely describe which orders
+are preserved by the OCaml program. The compiler and the hardware are free to
+optimize the program as long as they respect the ordering guarantees of the
+memory model. While programming directly under the relaxed memory model is
+difficult, the memory model also describes the conditions under which a program
+will only exhibit sequentially consistent behaviours. This guarantee is known
+as \emph{data race freedom implies sequential consistency} (DRF-SC). In this
+section, we shall describe this guarantee. In order to do this, we first need a
+number of definitions.
+
+\subsection{s:atomics}{Memory locations}
+
+OCaml classifies memory locations into \emph{atomic} and \emph{non-atomic}
+locations. Reference cells, array fields and mutable record fields are
+non-atomic memory locations. Immutable objects are non-atomic locations with an
+initialising write but no further updates. Atomic memory locations are those
+that are created using the \stdmoduleref{Atomic} module.
+
+\subsection{s:happens_before}{Happens-before relation}
+
+Let us imagine that the OCaml programs are executed by an abstract machine that
+executes one action at a time, arbitrarily picking one of the available domains
+at each step. We classify actions into two: \emph{inter-domain} and
+\emph{intra-domain}. An inter-domain action is one which can be observed and be
+influenced by actions on other domains. There are several inter-domain actions:
+
+\begin{itemize}
+  \item Reads and writes of atomic and non-atomic locations.
+  \item Spawn and join of domains.
+  \item Operations on mutexes.
+\end{itemize}
+
+\emph{(TODO: Include semaphores and condition variables in the inter-domain
+actions.)}
+
+On the other hand, intra-domain actions can neither be observed nor influence
+the execution of other domains. Examples include evaluating an arithmetic
+expression, calling a function, etc. The memory model specification ignores
+such intra-domain actions. In the sequel, we use the term action to indicate
+inter-domain actions.
+
+A totally ordered list of actions executed by the abstract machine is called an
+\emph{execution trace}. There might be several possible execution traces for a
+given program due to non-determinism.
+
+For a given execution trace, we define an irreflexive, transitive
+\emph{happens-before relation} that captures the causality between actions in
+the OCaml program. The happens-before relation is defined as follows:
+
+\begin{itemize}
+  \item We define the order in which a domain executes its actions as the
+    \emph{program order}. If an action "x" precedes another action "y" in
+    program order, then "x" precedes "y" in happens-before order.
+	\item If "x" is an initialising write to a memory location (atomic or
+		otherwise) and "y" is any other read or write to that memory location, then
+		"x" precedes "y" in happens-before order.
```

Comment:
Does this actually creates a full HB relationship? That seems really too strong to me. This essentially means that one can perform synchronisation without using any Atomic access or synchronisation library.

Consider this program:

```
let r = ref (ref 0)
let x = ref 0

let thread_a () =
    x := 1;
    r := ref 1

let thread_b () =
    let a = 2*!x in
    while !!r = 0 do () done;
    let b = 2*!x in
    (a, b)

let _ =
    Domain.create thread_b ();
    thread_a ();
    let (a, b) = Domain.join ();
    Printf.printf "%d %d" a b
```

If the above is correct, then there will be a HB relationship between the second read of x in thread_b and the write of x in thread_a. So the possible outcomes in the proposed memory model are (0, 2) and (2, 2). But sice we want to allow CSE between the definitions of a and b, we want to also allow (0, 0).

To me, the solution to fix this is simple: remove these buggy HB relationships, and exclude initialisations from conflicting accesses. As far as I can tel the DRF-SC theorem still hold.

Reply:
I have some thoughts on this point. I've pinged @stedolan to discuss this on a different channel to discuss this point.

Reply:
Had a chat with @stedolan on this issue and he agrees that it is incorrect to have added initialising write to the HB relationship. I'll remove initialising writes from HB and change the examples appropriately.

Reply:
Removed.